crypto/bio/bss_conn.c

```diff
@@ -202,15 +198,27 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
                 ERR_raise_data(ERR_LIB_SYS, i,
                                "calling connect(%s, %s)",
                                 c->param_hostname, c->param_service);
-                BIOerr(BIO_F_CONN_STATE, BIO_R_NBIO_CONNECT_ERROR);
+                if (i == ETIMEDOUT)
+                    BIOerr(BIO_F_CONN_STATE, BIO_R_NBIO_CONNECT_TIMEOUT);
+                else
+                    BIOerr(BIO_F_CONN_STATE, BIO_R_NBIO_CONNECT_ERROR);
                 ret = 0;
                 goto exit_loop;
             } else
                 c->state = BIO_CONN_S_OK;
             break;
 
         case BIO_CONN_S_CONNECT_ERROR:
-            BIOerr(BIO_F_CONN_STATE, BIO_R_CONNECT_ERROR);
+            i = get_last_socket_error();
```

Comment:
the callback is called, and could do anything with the errno.

Reply:
Sure a callback could do all kinds of things, but it should mess with `errno` only if there is good reason.
And how else would you suggest that one can detect the `getaddrinfo: Temporary failure in name resolution` situation?

Reply:
Yes, the getaddrinfo is something I hadn't considered before.
I don't really know, but wouldn't it be possible to get the 
errno value back what was entered to the error queue, right
when the getaddrinfo failed?

But I don't fully agree that this change is an improvement,
the difference when a connect returns TIMEOUT and when
it returns a different error, is not very big.  Just an lost
packet for instance.  But if a server is not online, you can
also try again it might just starting.
For the do_connect_retry it would be okay to retry as long
as the time is over.  I had not considered that possibility
but that can be done, but then I would just do that.

Reply:
> Yes, the getaddrinfo is something I hadn't considered before.
> I don't really know, but wouldn't it be possible to get the
> errno value back what was entered to the error queue, right
> when the getaddrinfo failed?

Indeed it would be possible and likely more safe to consult `errno` just after `getaddrinfo` failed, and then add a to-the-point entry/reason to the error queue

> But I don't fully agree that this change is an improvement,
> the difference when a connect returns TIMEOUT and when
> it returns a different error, is not very big. Just an lost
> packet for instance. But if a server is not online, you can
> also try again it might just starting.
> For the do_connect_retry it would be okay to retry as long
> as the time is over. I had not considered that possibility
> but that can be done, but then I would just do that.

You mean: as long the time is *not* over.
Good point, so we could at least to some extent lump timeout and error.
For some cases (`ETIMEDOUT`) I found that we need to call `BIO_reset()` to prevent immediate failure of the retry, but maybe such a call does not hurt also for all other error cases? Then much of the distinctions I added could be removed.

Reply:
> Indeed it would be possible and likely more safe to consult errno just after getaddrinfo failed, and then add a to-the-point entry/reason to the error queue

I really don't know but isn't the errno here entered into the error queue?
https://github.com/openssl/openssl/blob/a5a87011baeef71c86938a2bae54f89fbe99e5dc/crypto/bio/b_addr.c#L705-L708

It would feel much safer to get the errno back from the error queue,
but if that is not possible, if the DNS says no this name is not known *now*
you could as well try again after the short nap.

It would this connect_retry less surprising when there function simply retries
what it is doing until that timeout is over or it succeeds then it leaves the loop.

>  I found that we need to call BIO_reset() 

I think I begin to understand what this does.  in the case when
the connect fails, the socket is *not* closed, just the *connect*
failed.  and the state stays at BIO_CONN_S_CONNECT_ERROR.
Without that reset the state machine will not continue with
BIO_CONN_S_BEFORE.

Yes, if that is the correct behaviour for your application you just
can reset the socket and try to connect until that time is over.
I just think that this may be a parameter, since some clients,
may just want to connect once, and let the user decide what to
do next.

Reply:
> I really don't know but isn't the errno here entered into the error queue?

Yes, it is entered by `ERR_raise_data(ERR_LIB_SYS, get_last_socket_error(), "calling getaddrinfo()");` but then it is buried by `BIOerr(BIO_F_BIO_LOOKUP_EX, ERR_R_SYS_LIB);`.
Meanwhile `BIO_do_connect_retry()`, as you suggested, does not any more depend on `errno`.

Reply:
> Yes, if that is the correct behaviour for your application you just
can reset the socket and try to connect until that time is over.

This is meanwhile implemented.

Reply:
> I just think that this may be a parameter, since some clients,
may just want to connect once, and let the user decide what to
do next.

I like the idea to offer this option, although in this case the function degenerates to a normal `BIO_do_connect() call`.
Nevertheless I have implemented it. I did so not as an extra parameter but in case `timeout < 0`, which so far had no real use for `BIO_do_connect_retry()`.
This adds to uniformity and even simplifies the implementation.

Reply:
Yes.  I think that is getting better now.